pkg/tcl/testworkflowstcl/testworkflowcontroller/logs.go

```diff
@@ -0,0 +1,302 @@
+// Copyright 2024 Testkube.
+//
+// Licensed as a Testkube Pro file under the Testkube Community
+// License (the "License"); you may not use this file except in compliance with
+// the License. You may obtain a copy of the License at
+//
+//	https://github.com/kubeshop/testkube/blob/main/licenses/TCL.txt
+
+package testworkflowcontroller
+
+import (
+	"bufio"
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"regexp"
+	"strconv"
+	"strings"
+	"time"
+
+	errors2 "github.com/pkg/errors"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/client-go/kubernetes"
+
+	"github.com/kubeshop/testkube/cmd/tcl/testworkflow-init/data"
+	"github.com/kubeshop/testkube/pkg/api/v1/testkube"
+	"github.com/kubeshop/testkube/pkg/utils"
+)
+
+type Instruction struct {
+	Ref   string
+	Name  string
+	Value interface{}
+}
+
+func (i *Instruction) ToInternal() *testkube.TestWorkflowOutput {
+	if i == nil {
+		return nil
+	}
+	value := &i.Value
+	if i.Value == nil {
+		value = nil
+	}
+	return &testkube.TestWorkflowOutput{
+		Ref:   i.Ref,
+		Name:  i.Name,
+		Value: value,
+	}
+}
+
+type Comment struct {
+	Time   time.Time
+	Hint   *Instruction
+	Output *Instruction
+}
+
+type ContainerLog struct {
+	Time   time.Time
+	Log    []byte
+	Hint   *Instruction
+	Output *Instruction
+}
+
+type ContainerResult struct {
+	Status     testkube.TestWorkflowStepStatus
+	ExitCode   int
+	FinishedAt time.Time
+}
+
+var UnknownContainerResult = ContainerResult{
+	Status:   testkube.ABORTED_TestWorkflowStepStatus,
+	ExitCode: -1,
+}
+
+func GetContainerResult(c corev1.ContainerStatus) ContainerResult {
+	if c.State.Waiting != nil {
+		return ContainerResult{Status: testkube.QUEUED_TestWorkflowStepStatus, ExitCode: -1}
+	}
+	if c.State.Running != nil {
+		return ContainerResult{Status: testkube.RUNNING_TestWorkflowStepStatus, ExitCode: -1}
+	}
+	re := regexp.MustCompile(`^([^,]*),(0|[1-9]\d*)$`)
+	msg := c.State.Terminated.Message
+	match := re.FindStringSubmatch(msg)
+	if match == nil {
+		return ContainerResult{Status: testkube.ABORTED_TestWorkflowStepStatus, ExitCode: -1, FinishedAt: c.State.Terminated.FinishedAt.Time}
+	}
+	status := testkube.TestWorkflowStepStatus(match[1])
+	exitCode, _ := strconv.Atoi(match[2])
```

Comment:
is 3 elements guaranteed?

Reply:
It's based on `^([^,]*),(0|[1-9]\d*)$` regex above, so it will either be `nil`, or will have 3 groups (`0` - full match, `1` - the status from beginning of the string, `2` - the exit code from end of the string)